### DISCUSSION TITLE
how to modify name of table on the fly on a per-query basis?

### DISCUSSION THREAD

**Discussion by aradyaron** [sqlalchemy/sqlalchemy#8916] [Usage Questions]
### Describe the use case

In some database (like Oracle, Postgres) there is an option to partition/sharding tables.
This means the general database behavior and fields remain (Same model).
However you still can query on the specific sub partition sub part of the table, this is by replacing the tables name with the specific sub part name (for example 'sessions' -> 'sessions_2022_2023')

There is no easy way to do it currently.

Note that knowing the exact sub partition may occur in runtime, so creating a class representing the sub partition is not an option.

This should be controllable on the level of Session/Connection/Engine.

For example, I found a way to bypass this by addng an event on the level of the specific connection:

```
event.listen(connection, 'before_cursor_execute', create_replace_table_partition_name_hook(sub_partition_replacement),
                             retval=True)
```

where create_replace_table_partition_name_hook is

```
def create_replace_table_partition_name_hook(sub_partition_replacement: SubPartitionReplacement):

    # pylint: disable=unused-argument, too-many-arguments
    def replace_table_name(
        conn,
        cursor,
        statement,
        parameters,
        context,
        executemany,
    ):
        logger: Logger = get_logger()
        logger.debug(
            'Inside replace table partition name hook',
            main_table_name=sub_partition_replacement.main_table_name,
            partition_table_name=sub_partition_replacement.partition_table_name,
        )
        # avoid double call
        if not sub_partition_replacement.partition_table_name in statement:
            # This hook replaces all instances of the main table on the statement (recognized by .TABLE)
            # with the name of the sub partition of the same table.
            # This allows limiting the scope of the statement to the same sub partition.
            statement = statement.replace(
                f'.{sub_partition_replacement.main_table_name}',
                f'.{sub_partition_replacement.partition_table_name}',
            )
        return statement, parameters

    return replace_table_name
```

### Databases / Backends / Drivers targeted

General

### Example Use

```
event.listen(connection, 'before_cursor_execute', create_replace_table_partition_name_hook(sub_partition_replacement),
                             retval=True)
```

where create_replace_table_partition_name_hook is

```
def create_replace_table_partition_name_hook(sub_partition_replacement: SubPartitionReplacement):

    # pylint: disable=unused-argument, too-many-arguments
    def replace_table_name(
        conn,
        cursor,
        statement,
        parameters,
        context,
        executemany,
    ):
        logger: Logger = get_logger()
        logger.debug(
            'Inside replace table partition name hook',
            main_table_name=sub_partition_replacement.main_table_name,
            partition_table_name=sub_partition_replacement.partition_table_name,
        )
        # avoid double call
        if not sub_partition_replacement.partition_table_name in statement:
            # This hook replaces all instances of the main table on the statement (recognized by .TABLE)
            # with the name of the sub partition of the same table.
            # This allows limiting the scope of the statement to the same sub partition.
            statement = statement.replace(
                f'.{sub_partition_replacement.main_table_name}',
                f'.{sub_partition_replacement.partition_table_name}',
            )
        return statement, parameters

    return replace_table_name
```

**Comment by zzzeek** [comment]
There's a recipe for the exact use case on the wiki at : https://github.com/sqlalchemy/sqlalchemy/wiki/SessionModifiedSQL   

Ideally the same recipe can be used with the `.execution_options()` feature, though the example as given there shows the flush process being affected too where we don't yet have a way to affect execution options, so the above wiki entry is currently the most general purpose approach for now.

**Reply by aradyaron** [reply]
This approach indeed works, however this utilizes string manipulation, rather than the direct approach of changing the table name.
Explicitly changing table name will be less error prone, and more maintainable as well.

**Reply by zzzeek** [reply]
what do we mean by "change the table name" ?   if you have a `Table` object, that's not a statement local object.  It's global (and also heavily embedded within cached constructs).  so you would not want to do something like `table.name = newname`.  that name should be considered as immutable.

you can make new Table objects and alias to them, if that's what you want to do, this would be fine if you have a limited number of "shard" tables, and not like, millions:

```py
my_alt_table = Table("sessions_2022_2023", metadata, ...columns)

SessionsAlt = aliased(Sessions, my_alt_table.alias())

data = session.query(SessionsAlt).filter(SessionsAlt.col == 'x')...

```

however this is not too different from the subclasses-per-table approach , where you can indeed make new subclasses on the fly and look them up in a cache so that you don't have to keep recreating them.    The subclasses-per-table approach is illustrated at https://github.com/sqlalchemy/sqlalchemy/wiki/EntityName.

**Reply by aradyaron** [reply]
In that case how developing using what can be called an 'aliased' name? and either enforce overriding the name/ have a default table name?
marking a table name as aliased, and allowing (on the level of a specific instance of session/connection/engine) to control it?
Please note that I am suggesting general ideas, however I am not familiar enough with sqlalchemy to fully understand the implications.

**Reply by zzzeek** [reply]
one could make a custom FromClause implementation to do this, i dont have time to illustrate one right now however it would be similar to what's in the [view recipe](https://github.com/sqlalchemy/sqlalchemy/wiki/Views).  

What this means though is that each time you query with a different "shard", a brand new entry gets put into the [statement cache](https://docs.sqlalchemy.org/en/14/core/connections.html#sql-compilation-caching).   This is why for scalability purposes we tend to favor late statement rewriting, which is also how the [schema translate](https://docs.sqlalchemy.org/en/14/core/connections.html#translation-of-schema-names) feature works.
